File: ios/Classes/NativeEncryptorPlugin.swift
import Flutter
import UIKit

public class NativeEncryptorPlugin: NSObject, FlutterPlugin {
    private let aesEncryptionService = AESEncryptionService()

    public static func register(with registrar: FlutterPluginRegistrar) {
        let channel = FlutterMethodChannel(name: "native_encryptor", binaryMessenger: registrar.messenger())
        let instance = NativeEncryptorPlugin()
        registrar.addMethodCallDelegate(instance, channel: channel)
    }

    public func handle(_ call: FlutterMethodCall, result: @escaping FlutterResult) {
        switch call.method {
        case "getPlatformVersion":
            result("iOS " + UIDevice.current.systemVersion)
        case "encrypt":
            handleEncrypt(call: call, result: result)
        case "decrypt":
            handleDecrypt(call: call, result: result)
        case "encrypt_with_custom_iv":
            handleEncryptWithCustomIV(call: call, result: result)
        case "decrypt_with_custom_iv":
            handleDecryptWithCustomIV(call: call, result: result)
        default:
            result(FlutterMethodNotImplemented)
        }
    }

    private func handleEncrypt(call: FlutterMethodCall, result: @escaping FlutterResult) {
        guard
            let args = call.arguments as? [String: String],
            let passPhrase = args["passPhrase"],
            let contentToEncrypt = args["contentToEncrypt"]
        else {
            result(FlutterError(code: "INVALID_ARGUMENTS", message: "Missing encryption arguments", details: nil))
            return
        }

        do {
            let encryptedValue = try aesEncryptionService.encrypt(passphrase: passPhrase, plainText: contentToEncrypt)
            result(encryptedValue)
        } catch let error {
            result(FlutterError(code: "ENCRYPTION_ERROR", message: error.localizedDescription, details: nil))
        }
    }

    private func handleDecrypt(call: FlutterMethodCall, result: @escaping FlutterResult) {
        guard
            let args = call.arguments as? [String: String],
            let passPhrase = args["passPhrase"],
            let concatenatedCipherText = args["concatenatedCipherText"]
        else {
            result(FlutterError(code: "INVALID_ARGUMENTS", message: "Missing decryption arguments", details: nil))
            return
        }

        do {
            let decryptedValue = try aesEncryptionService.decrypt(
                passphrase: passPhrase,
                concatenatedCipherText: concatenatedCipherText
            )
            result(decryptedValue)
        } catch let error {
            result(FlutterError(code: "DECRYPTION_ERROR", message: error.localizedDescription, details: nil))
        }
    }

    private func handleEncryptWithCustomIV(call: FlutterMethodCall, result: @escaping FlutterResult) {
        guard
            let args = call.arguments as? [String: String],
            let passPhrase = args["passPhrase"],
            let iv = args["iv"],
            let contentToEncrypt = args["contentToEncrypt"]
        else {
            result(FlutterError(code: "INVALID_ARGUMENTS", message: "Missing custom IV encryption arguments", details: nil))
            return
        }

        if let encryptedValue = CustomIVEDService.encrypt(passPhrase: passPhrase, iv: iv, plainText: contentToEncrypt) {
            result(encryptedValue)
        } else {
            result(FlutterError(code: "ENCRYPTION_ERROR", message: "Custom IV encryption failed", details: nil))
        }
    }

    private func handleDecryptWithCustomIV(call: FlutterMethodCall, result: @escaping FlutterResult) {
        guard
            let args = call.arguments as? [String: String],
            let passPhrase = args["passPhrase"],
            let iv = args["iv"],
            let concatenatedCipherText = args["concatenatedCipherText"]
        else {
            result(FlutterError(code: "INVALID_ARGUMENTS", message: "Missing custom IV decryption arguments", details: nil))
            return
        }

        if let decryptedValue = CustomIVEDService.decrypt(passPhrase: passPhrase, iv: iv, cipherText: concatenatedCipherText) {
            result(decryptedValue)
        } else {
            result(FlutterError(code: "DECRYPTION_ERROR", message: "Custom IV decryption failed", details: nil))
        }
    }
}

File: android/src/main/java/com/example/native_encryptor/NativeEncryptorPlugin.java
package com.example.native_encryptor;

import android.os.Build;

import androidx.annotation.NonNull;
import androidx.annotation.RequiresApi;

import java.util.HashMap;
import java.util.Objects;

import io.flutter.embedding.engine.plugins.FlutterPlugin;
import io.flutter.plugin.common.MethodCall;
import io.flutter.plugin.common.MethodChannel;
import io.flutter.plugin.common.MethodChannel.MethodCallHandler;
import io.flutter.plugin.common.MethodChannel.Result;

/**
 * NativeEncryptorPlugin
 */
public class NativeEncryptorPlugin implements FlutterPlugin, MethodCallHandler {
    /// The MethodChannel that will the communication between Flutter and native Android
    ///
    /// This local reference serves to register the plugin with the Flutter Engine and unregister it
    /// when the Flutter Engine is detached from the Activity
    private MethodChannel channel;
    AESEncryptionService aesEncryptionService = new AESEncryptionService();

    @Override
    public void onAttachedToEngine(@NonNull FlutterPluginBinding flutterPluginBinding) {
        channel = new MethodChannel(flutterPluginBinding.getBinaryMessenger(), "native_encryptor");
        channel.setMethodCallHandler(this);
    }

    @Override
    public void onMethodCall(@NonNull MethodCall call, @NonNull Result result) {
        if (call.method.equals("getPlatformVersion")) {
            result.success("Android " + android.os.Build.VERSION.RELEASE);
        } else if (call.method.equals("encrypt")) {
            try {
                HashMap<String, String> data = (HashMap<String, String>) call.arguments;
                PlugInUtility.logInfo("Received params - " + data.toString());
                String encryptedValue = onEncryptMethodCall(Objects.requireNonNull(data.get("passPhrase")), Objects.requireNonNull(data.get("contentToEncrypt")));
                PlugInUtility.logInfo("Encrypted data - " + encryptedValue);
                result.success(encryptedValue);
            } catch (Exception e) {
                PlugInUtility.logError("Error on encryption - " + e.getMessage(), e.fillInStackTrace());
                result.error("", e.getMessage(), e.getLocalizedMessage());
            }
        } else if (call.method.equals("decrypt")) {
            try {
                HashMap<String, String> data = (HashMap<String, String>) call.arguments;
                PlugInUtility.logInfo("Received params - " + data.toString());
                String decryptedValue = onDecryptMethodCall(Objects.requireNonNull(data.get("passPhrase")), Objects.requireNonNull(data.get("concatenatedCipherText")));
                PlugInUtility.logInfo("Decrypted data - " + decryptedValue);
                result.success(decryptedValue);
            } catch (Exception e) {
                PlugInUtility.logError("Error on decryption - " + e.getMessage(), e.fillInStackTrace());
                result.error("", e.getMessage(), e.getLocalizedMessage());
            }
        }

        /// encrypt/decrypt with custom IV
        else if (call.method.equals("encrypt_with_custom_iv")) {
            try {
                PlugInUtility.logInfo("encrypt_with_custom_iv in action....");
                HashMap<String, String> data = (HashMap<String, String>) call.arguments;
                PlugInUtility.logInfo("Received params - " + data.toString());
                String encryptedValue = onEncryptWithIVMethodCall(Objects.requireNonNull(data.get("passPhrase")), Objects.requireNonNull(data.get("iv")), Objects.requireNonNull(data.get("contentToEncrypt")));
                PlugInUtility.logInfo("Encrypted data - " + encryptedValue);
                result.success(encryptedValue);
            } catch (Exception e) {
                PlugInUtility.logError("Error on encryption - " + e.getMessage(), e.fillInStackTrace());
                result.error("", e.getMessage(), e.getLocalizedMessage());
            }
        } else if (call.method.equals("decrypt_with_custom_iv")) {
            try {
                PlugInUtility.logInfo("decrypt_with_custom_iv in action....");
                HashMap<String, String> data = (HashMap<String, String>) call.arguments;
                PlugInUtility.logInfo("Received params - " + data.toString());
                String decryptedValue = onDecryptWithIVMethodCall(Objects.requireNonNull(data.get("passPhrase")), Objects.requireNonNull(data.get("iv")), Objects.requireNonNull(data.get("concatenatedCipherText")));
                PlugInUtility.logInfo("Decrypted data - " + decryptedValue);
                result.success(decryptedValue);
            } catch (Exception e) {
                PlugInUtility.logError("Error on decryption - " + e.getMessage(), e.fillInStackTrace());
                result.error("", e.getMessage(), e.getLocalizedMessage());
            }
        } else {
            result.notImplemented();
        }
    }

    String onEncryptMethodCall(@NonNull String passPhrase, @NonNull String content) {
        return aesEncryptionService.encrypt(passPhrase, content);
    }

    String onDecryptMethodCall(@NonNull String passPhrase, @NonNull String encryptedValue) {
        return aesEncryptionService.decrypt(passPhrase, encryptedValue);
    }

    @RequiresApi(api = Build.VERSION_CODES.O)
    String onEncryptWithIVMethodCall(@NonNull String passPhrase, @NonNull String iv, @NonNull String content) {
        return CustomIVEDService.getEncryptedString(passPhrase, iv, content);
    }

    @RequiresApi(api = Build.VERSION_CODES.O)
    String onDecryptWithIVMethodCall(@NonNull String passPhrase, @NonNull String iv, @NonNull String encryptedValue) {
        return CustomIVEDService.getDecryptedString(passPhrase, iv, encryptedValue);
    }

    @Override
    public void onDetachedFromEngine(@NonNull FlutterPluginBinding binding) {
        channel.setMethodCallHandler(null);
    }
}
